### ISSUE TITLE
In TypeScript multi-line type definitions, Prettier should put the semicolon on a separate line

### ISSUE THREAD

**Issue by Zamiell** [prettier/prettier#9370]
Consider the following TypeScript type definition, formatted by Prettier:

```
type MyType =
  | "Foo1"
  | "Foo2"
  | "Foo3"
  | "Foo4"
  | "Foo5;"
```

Now, what if I need to add a new line to my type definition? I'll just quickly add "Foo6":

```
type MyType =
  | "Foo1"
  | "Foo2"
  | "Foo3"
  | "Foo4"
  | "Foo5";
  | "Foo6"; // Oops - now the code refuses to compile because of the double semicolons
```

Even after saving the file / formatting it with Prettier, the semicolon remains on the "Foo5" line. So now I have to go back and manually delete the semicolon from that line. Annoying!

From what I can tell, a more "correct" way to format this code snippet would be to put the semicolon on a new line.
Consider the following code snippet [from the official TypeScript compiler](https://github.com/microsoft/TypeScript/blob/master/src/compiler/types.ts#L449-L456):

```
    export type TriviaSyntaxKind =
        | SyntaxKind.SingleLineCommentTrivia
        | SyntaxKind.MultiLineCommentTrivia
        | SyntaxKind.NewLineTrivia
        | SyntaxKind.WhitespaceTrivia
        | SyntaxKind.ShebangTrivia
        | SyntaxKind.ConflictMarkerTrivia
        ;
```

This style seems be be used throughout the repository (correct me if I am wrong).

Unfortunately, if I want to use this style, after sticking the semicolon on a new line, Prettier will automatically move it back to the end of the previous line (as one would expect).

I propose that Prettier should adopt the official TypeScript style for any multi-line type definition - by default, without adding an extra option. Since I presume that this would be a breaking change, it should be a candidate for the 3.0 release.

Thanks for the excellent formatter.

**Comment by evilebottnawi** [comment]
It is expected, we don't want to change output, this styles does not lead to improved reading in many cases, only for simple case what you provided

**Comment by Zamiell** [comment]
> this styles does not lead to improved reading in many cases

In which cases?

**Comment by thorn0** [comment]
The proposed formatting makes sense indeed as it's more diff-friendly than the current one. However, it feels unusual and inconsistent as there is no other cases where Prettier prints such a dangling semicolon. That's why here is what I'd like to propose instead:

```ts
    export type TriviaSyntaxKind = (
        | SyntaxKind.SingleLineCommentTrivia
        | SyntaxKind.MultiLineCommentTrivia
        | SyntaxKind.NewLineTrivia
        | SyntaxKind.WhitespaceTrivia
        | SyntaxKind.ShebangTrivia
        | SyntaxKind.ConflictMarkerTrivia
    );
```

**Comment by Zamiell** [comment]
That seems like a reasonable proposal. Doing that would make the formatting look similar to a multi-line conditional statement, e.g.:

```typescript
// An example mock multi-line conditional statement
// Note that in Prettier 3.0, operators will be put at the beginning of the line, not the end
const isTriviaSyntax = (
  trivia === SyntaxKind.SingleLineCommentTrivia
  || trivia === SyntaxKind.MultiLineCommentTrivia
  || trivia === SyntaxKind.NewLineTrivia
  || trivia === SyntaxKind.WhitespaceTrivia
  || trivia === SyntaxKind.ShebangTrivia
  || trivia === SyntaxKind.ConflictMarkerTrivia
);
```

And yes, thank you, I forgot to include the argument about making code more diff-friendly.
This kind of change has all the same benefits that trailing commas do, AFAIK.

**Comment by Zamiell** [comment]
> Doing that would make the formatting look similar to a multi-line conditional statement,

Actually, I was mistaken. Prettier will actually format what I wrote to:

```typescript
const isTriviaSyntax =
    || trivia === SyntaxKind.SingleLineCommentTrivia
    || trivia === SyntaxKind.MultiLineCommentTrivia
    || trivia === SyntaxKind.NewLineTrivia
    || trivia === SyntaxKind.WhitespaceTrivia
    || trivia === SyntaxKind.ShebangTrivia
    || trivia === SyntaxKind.ConflictMarkerTrivia;
```

Which seems ugly and wrong, for the same reasons that thorn0's proposal is reasonable. 😆

**Comment by Zamiell** [comment]
In conclusion, I would expect thorn0's proposal to format both of these code snippets in the same way (e.g. with a parenthesis) - not just the TypeScript-specific-type-definition code snippet.

(I am for this proposal, but am just pointing this out.)

**Comment by thorn0** [comment]
related: https://mobile.twitter.com/kuvos/status/1336615168223154176

**Comment by Zamiell** [comment]
@thorn0 Is there a good way to generate more discussion on this proposal? Is Prettier accepting PRs for this?

**Comment by thorn0** [comment]
@Zamiell Yes, a PR implementing https://github.com/prettier/prettier/issues/9370#issuecomment-717857062 would be welcome. Please include the code from the tweet above in tests.

**Comment by thorn0** [comment]
The changelog entry in the PR needs to include a paragraph of text for the release blog post. In that paragraph, we need to sell this change to the users. It's quite a drastic change leading to massive diffs, so if the text isn't be persuasive enough, we might end up restricting this change only to the positions where the current formatting for multiline union types looks especially bad (e.g., function return types).

Another thing to keep in mind, is the `semi: false` options. The current formatting for type aliases, interface entries, etc. should stay as it is now in that mode, but union types used as function return types (and probably in other places) still should need parens.
